clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -76,41 +75,65 @@
+    public SaslClientAuthenticator(String node, Subject subject, String servicePrincipal, String host, KerberosNameParser kerberosNameParser, SaslMechanism mechanism) throws IOException {
         this.node = node;
         this.subject = subject;
         this.host = host;
         this.servicePrincipal = servicePrincipal;
         this.kerberosNameParser = kerberosNameParser;
+        this.mechanism = mechanism;
     }
 
     public void configure(TransportLayer transportLayer, PrincipalBuilder principalBuilder, Map<String, ?> configs) throws KafkaException {
         try {
             this.transportLayer = transportLayer;
-
-            // determine client principal from subject.
-            Principal clientPrincipal = subject.getPrincipals().iterator().next();
-            this.clientPrincipalName = kerberosNameParser.parse(clientPrincipal.getName()).toString();
-            this.saslClient = createSaslClient();
+            switch (mechanism) {
+                case GSSAPI:
+                    this.saslClient = createSaslKerberosClient();
+                    break;
+                case PLAIN:
+                    this.saslClient = createSaslPlainClient();
+                    break;
+                default:
+                    throw new KafkaException("Unsupported SASL mechanism " + mechanism);
+            }
         } catch (Exception e) {
             throw new KafkaException("Failed to configure SaslClientAuthenticator", e);
         }
     }
 
-    private SaslClient createSaslClient() {
+    private SaslClient createSaslKerberosClient() {
         try {
+            // determine client principal from subject.
+            Principal clientPrincipal = subject.getPrincipals().iterator().next();
+            this.clientPrincipalName = kerberosNameParser.parse(clientPrincipal.getName()).toString();
             return Subject.doAs(subject, new PrivilegedExceptionAction<SaslClient>() {
                 public SaslClient run() throws SaslException {
-                    String[] mechs = {"GSSAPI"};
+                    String[] mechs = {mechanism.mechanismName()};
                     LOG.debug("Creating SaslClient: client={};service={};serviceHostname={};mechs={}",
                         clientPrincipalName, servicePrincipal, host, Arrays.toString(mechs));
                     return Sasl.createSaslClient(mechs, clientPrincipalName, servicePrincipal, host, null,
-                            new ClientCallbackHandler());
+                            new ClientCallbackHandler(subject, mechanism));
                 }
             });
         } catch (Exception e) {
-            throw new KafkaException("Failed to create SASL client", e);
+            throw new KafkaException("Failed to create SASL/Kerberos client", e);
+        }
+    }
+    
+    private SaslClient createSaslPlainClient() {
```

Comment:
It seems to me that we could have a single `createSaslClient`. Most of the code is the same apart from the mechanism, clientPrincipal and the exception message.